File: Library/Classes/Core/Classes/UI/CSSelectNameController.m
//
// Created by Rene Dohan on 26/05/18.
// Copyright (c) 2018 Renetik Software. All rights reserved.
//

#import "CSSelectNameController.h"
#import "CSMenuItem.h"
#import "CSSearchBarController.h"
#import "CSListData.h"
#import "CSResponse.h"

@interface CSSelectNameController () <UITableViewDelegate, UITableViewDataSource>
@property (nonatomic, strong)  NSMutableArray<CSName *> * names;
@end

@implementation CSSelectNameController {
    NSArray<CSName *> *_filteredData;
    void (^_onSelected)(CSName *);
    CSResponse *(^_onDelete)(CSName *);
    NSString *_deleteTitle;
}

- (instancetype)constructByData:(NSObject <CSListData> *)names :(void (^)(CSName *))onSelected {
    return [self constructByNames:names.list :onSelected];
}

- (instancetype)constructByData:(NSObject <CSListData> *)names :(void (^)(CSName *))onSelected
        :(NSString *)deleteTitle :(CSResponse *(^)(CSName *))onDelete {
    [self constructByData:names :onSelected];
    _onDelete = [onDelete copy];
    _deleteTitle = deleteTitle;
    [[self menuItem:nil type:UIBarButtonSystemItemEdit] onClick:^(CSMenuItem *item) {
        item.systemItem = _table.toggleEditing.editing ? UIBarButtonSystemItemCancel : UIBarButtonSystemItemEdit;
    }];
    return self;
}

- (instancetype)constructByNames:(NSArray<CSName *> *)names :(void (^)(CSName *))onSelected {
    _names = names.mutableCopy;
    _onSelected = [onSelected copy];
    _filteredData = [NSArray arrayWithArray:_names];
    return self;
}

- (instancetype)constructByNames:(NSArray<CSName *> *)names :(void (^)(CSName *))onSelected :(NSString *)clearTitle {
    [self constructByNames:names :onSelected];
    wlet _self = self;
    [self menuItem:clearTitle :^(CSMenuItem *item) {[_self onClearClick];}];
    return self;
}

- (instancetype)constructByStrings:(NSArray<NSString *> *)strings :(void (^)(NSNumber *))onSelected :(NSString *)clearTitle {
    [self constructByNames:[CSName createNamesFromStrings:strings] :^(CSName *name) {
        runWith(onSelected, name ? @(name.index) : nil);
    } :clearTitle];
    return self;
}

- (instancetype)constructByNames:(NSArray<CSName *> *)names :(void (^)(CSName *))onSelected
        :(NSString *)deleteTitle :(CSResponse * (^)(CSName *))onDelete {
    [self constructByNames:names :onSelected];
    _onDelete = [onDelete copy];
    _deleteTitle = deleteTitle;
    [[self menuItem:nil type:UIBarButtonSystemItemEdit] onClick:^(CSMenuItem *item) {
        item.systemItem = _table.toggleEditing.editing ? UIBarButtonSystemItemCancel : UIBarButtonSystemItemEdit;
    }];
    return self;
}

- (void)loadView {
    self.view = _table = [UITableView withSize:100 :100];
    _table.tableHeaderView = _search = [UISearchBar withSize:100 :40];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [[_table setupTable:self] hideEmptyCellSplitterBySettingEmptyFooter].allowsMultipleSelectionDuringEditing = NO;
    wlet _self = self;
    [CSSearchBarController.new construct:(self) :_search :^(NSString *searchText) {[_self reload];}];
}

- (void)onClearClick {
    [self.navigationController popViewController];
    runWith(_onSelected, _selectedRow = nil);
}

- (instancetype)setData:(NSArray<CSName *> *)names {
    _names = names;
    [_table reloadData];
    return self;
}

- (void)onFilterData:(NSString *)searchText {
    [self reload];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)path {
    UITableViewCell *cell = [tableView getCellWithStyle:@"Cell" :UITableViewCellStyleDefault];
    cell.textLabel.textColor = self.secondaryColor;
    cell.textLabel.text = [_filteredData at:path.index].name;
    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _filteredData.size;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)path {
    _selectedRow = [_filteredData at:path.index];
    [self.navigationController popViewController];
    runWith(_onSelected, _selectedRow);
}

- (nullable NSArray<UITableViewRowAction *> *)tableView:(UITableView *)tableView editActionsForRowAtIndexPath:(NSIndexPath *)path {
    if (_onDelete) {
        wlet _self = self;
        let delete = [UITableViewRowAction rowActionWithStyle:UITableViewRowActionStyleDestructive title:_deleteTitle handler
        :^(UITableViewRowAction *action, NSIndexPath *indexPath) {[_self onDeleteAction:path];}];
        delete.backgroundColor = self.primaryColor;
        return @[delete];
    }
    return nil;
}

- (void)onDeleteAction:(NSIndexPath *)path {
    let value = [_filteredData at:path.index];
    wlet _self = self;
    [_onDelete(value) onSuccess:^(id o) {
        [_self.names remove:value];
        [_self reload];
    }];
}

- (void)reload {
    _filteredData = [_names filterBySearch:_search.text];
    [_table reloadData];
}
@end

File: Renetik/Renetik/Classes/Core/Classes/Controller/CSSelectNameController.swift
//
//  CSSelectNameController.swift
//  Renetik
//
//  Created by Rene Dohan on 2/20/19.
//

import RenetikObjc

public class CSSelectNameController: CSMainController, UITableViewDelegate, UITableViewDataSource {
    public let table = UITableView.construct()
    public let search = CSSearchBarController()
    public var onCellCreate: ((UITableViewCell) -> Void)?
    public var selectedName: CSName?
    private var names: [CSName] = []
    private var filteredData: [CSName] = []
    private var onSelected: ((CSName) -> Void)!
    private var onDelete: ((CSName) -> CSResponseProtocol)?

    @discardableResult
    public func construct(data: [CSName], onSelected: @escaping (CSName) -> Void) -> Self {
        names = data
        self.onSelected = onSelected
        return self
    }

    @discardableResult
    public func addDelete<T: AnyObject>(_ onDelete: @escaping (CSName) -> CSResponse<T>) -> Self {
        self.onDelete = onDelete
        menu(type: .edit) { $0.systemItem = self.table.toggleEditing().isEditing ? .cancel : .edit }
        return self
    }

    override public func onViewWillAppear() {
        super.onViewWillAppear()
        view.add(view: table).matchParent()
        table.hideEmptyCellsSeparatorByEmptyFooter()
        table.allowsMultipleSelectionDuringEditing = false

        search.construct(by: self) { _ in self.reload() }
        table.tableHeaderView = search.bar

        table.delegates(self)
        reload()
    }

    private func reload() {
        filteredData = names.filter(bySearch: search.text)
        table.reload()
    }

    public func tableView(_ tableView: UITableView,
                          cellForRowAt path: IndexPath) -> UITableViewCell {
        tableView.cell(style: .default, onCreate: onCellCreate)
            .also { $0.textLabel!.text = filteredData[path.row].name }
    }

    public func tableView(_: UITableView,
                          numberOfRowsInSection _: Int) -> Int {
        filteredData.count
    }

    public func tableView(_: UITableView,
                          didSelectRowAt path: IndexPath) {
        selectedName = filteredData[path.row]
        navigation.popViewController()
        onSelected!(selectedName!)
    }

    public func tableView(_: UITableView,
                          canEditRowAt _: IndexPath) -> Bool {
        onDelete.notNil
    }

    public func tableView(_: UITableView,
                          commit editingStyle: UITableViewCell.EditingStyle,
                          forRowAt path: IndexPath) {
        if editingStyle == .delete {
            let value = filteredData[path.row]
            onDelete?(value).onSuccess {
                self.names.remove(value)
                if self.names.isEmpty { navigation.popViewController() } else { self.reload() }
            }
        }
    }
}
